src/client/render/pipeline.h

```diff
@@ -0,0 +1,391 @@
+	virtual u8 getTextureCount() override;
+
+	/**
+	 * Get a texture by index. 
+	 * Returns nullptr is the texture does not exist.
+	 */
+	virtual video::ITexture *getTexture(u8 index) override;
+private:
+	RenderSource *upstream { nullptr };
+};
+
+/**
+ * Implements direct output to screen framebuffer.
+ */
+class ScreenTarget : public RenderTarget
+{
+public:
+	virtual void activate(PipelineContext *context) override;
+	virtual void reset(PipelineContext *context) override;
+private:
+	core::dimension2du size;
+};
+
+class DynamicTarget : public RenderTarget
+{
+public:
+	bool isConfigured() { return upstream != nullptr; }
+	void setRenderTarget(RenderTarget *value) { upstream = value; }
+	virtual void activate(PipelineContext *context) override;
+private:
+	RenderTarget *upstream { nullptr };
+};
+
+/**
+ * Base class for rendering steps in the pipeline
+ */
+class RenderStep : virtual public RenderPipelineObject
+{
+public:
+	/**
+	 * Assigns render source to this step.
+	 * 
+	 * @param source source of rendering information
+	 */
+	virtual void setRenderSource(RenderSource *source) = 0;
+
+	/**
+	 * Assigned render target to this step.
+	 * 
+	 * @param target render target to send output to.
+	 */
+	virtual void setRenderTarget(RenderTarget *target) = 0;
+
+	/**
+	 * Runs the step. This method is invoked by the pipeline.
+	 */
+	virtual void run(PipelineContext *context) = 0;
+};
+
+/**
+ * Provides default empty implementation of supporting methods in a rendering step.
+ */
+class TrivialRenderStep : public RenderStep
+{
+public:
+	virtual void setRenderSource(RenderSource *source) override {}
+	virtual void setRenderTarget(RenderTarget *target) override {}
+	virtual void reset(PipelineContext *) override {}
+};
+
+/**
+ * Dynamically changes render target of another step.
+ * 
+ * This allows re-running parts of the pipeline with different outputs
+ */
+class SetRenderTargetStep : public TrivialRenderStep
+{
+public:
+	SetRenderTargetStep(RenderStep *step, RenderTarget *target);
+	virtual void run(PipelineContext *context) override;
+private:
+	RenderStep *step;
+	RenderTarget *target;
+};
+
+/**
+ * Render Pipeline provides a flexible way to execute rendering steps in the engine.
+ * 
+ * RenderPipeline also implements @see RenderStep, allowing for nesting of the pipelines.
+ */
+class RenderPipeline : public RenderStep
+{
+public:
+	/**
+	 * Add a step to the end of the pipeline
+	 * 
+	 * @param step reference to a @see RenderStep implementation.
+	 */
+	void addStep(RenderStep *step)
+	{
+		m_pipeline.push_back(step);
+	}
+
+	/**
+	 * Capture ownership of a dynamically created @see RenderStep instance.
+	 * 
+	 * RenderPipeline will delete the instance when the pipeline is destroyed.
+	 * 
+	 * @param step reference to the instance.
+	 * @return RenderStep* value of the 'step' parameter.
+	 */
+	template<class T>
+	T *own(T *object)
+	{
+		m_objects.push_back(std::unique_ptr<RenderPipelineObject>(object));
+		return object;
+	}
```

Comment:
This should probably take a `unique_ptr`:
```suggestion
	template<class T>
	T &own(std::unique_ptr<T> object)
	{
		T &ref = *object;
		m_objects.push_back(std::move(object));
		return ref;
	}
```
(Could still return a raw ptr, if you want.)

You could also add a helper member function for shortness:
```cpp
template<typename T, typename... Args>
T &make_owned(Args&&... args) {
	return own(std::make_unique<T>(std::forward<Args>(args)...));
}
```

Then you have:
```cpp
auto screen = pipeline->own(new ScreenTarget()); // old
auto &screen = pipeline->own(std::make_unique<ScreenTarget>()); // new
auto &screen = pipeline->make_owned<ScreenTarget>(); // shorter
```

Reply:
Added createOwned<T>() and addStep<T>(...) as a shortcut for addStep(createOwned<T>()).